File: Bingo/Bingo.Core/TicketBuilder.cs
namespace Bingo.Core
{

    class Column : List<int>
    {
        public Column(int index) : base(3)
        {
            Index = index;
        }
        public int Index { get; set; }
    }
    class TicketBuilder
    {
        public Ticket Build()
        {
            return new Ticket(Cells);
        }

        public int NumbersCount = 0;
        public Column[] Ranges = new Column[9];
        public int?[][] Cells { get; set; } = new int?[3][];
        public int Index { get; set; }
        public TicketBuilder(int index)
        {
            Index = index;
            for (int i = 0; i < 3; i++)
            {
                Cells[i] = new int?[9];
            }

            for (int i = 0; i < 9; i++)
            {
                Ranges[i] = new Column(i);
            }
        }

        internal void Place(int v, int rangeIndex)
        {
            NumbersCount++;
            Ranges[rangeIndex].Add(v);
        }

        internal bool CanAccept(int v, int rangeIndex)
        {
            var howManyInRange = Ranges[rangeIndex].Count;

            return howManyInRange < 3 && NumbersCount < 15;
        }

        internal void BalanceEmptyCells()
        {
            var random = new Random();
            var emptyCellsLeft = new Pair[] { new Pair(0, 4), new Pair(1, 4), new Pair(2, 4) };

            foreach (var range in Ranges.OrderBy(r => r.Count))
            {
                if (range.Count == 1)
                {
                    var rowindicesLeft = new List<int>() { 0, 1, 2 };
                    var ordo = emptyCellsLeft.OrderByDescending(x => x.Count).ToArray();
                    var row = ordo[0];
                    var columnIndex = range.Index;
                    rowindicesLeft.Remove(row.Index);
                    row.Count--;

                    row = ordo[1];

                    rowindicesLeft.Remove(row.Index);
                    row.Count--;

                    Cells[rowindicesLeft.First()][columnIndex] = range.First();
                }

                if (range.Count == 2)
                {
                    var rowindicesLeft = new List<int>() { 0, 1, 2 };
                    var row = emptyCellsLeft.OrderByDescending(x => x.Count).First();
                    var columnIndex = range.Index;

                    rowindicesLeft.Remove(row.Index);
                    row.Count--;

                    var ordered = range.Order().ToArray();
                    Cells[rowindicesLeft.First()][columnIndex] = ordered[0];
                    Cells[rowindicesLeft.Last()][columnIndex] = ordered[1];
                }

                if (range.Count == 3)
                {
                    var columnIndex = range.Index;

                    var ordered = range.Order().ToArray();

                    Cells[0][columnIndex] = ordered[0];
                    Cells[1][columnIndex] = ordered[1];
                    Cells[2][columnIndex] = ordered[2];
                }
            }
            //    "0|6|3",
            //    "1|4|4",
            //    "2|2|5",
            //    "3|0|6",
        }

        class Pair
        {
            public Pair(int index, int count)
            {
                Index = index;
                Count = count;
            }
            public int Index;
            public int Count;
        }
    }

}

File: core/src/main/kotlin/org/example/core/generation/TicketBuilder.kt
package org.example.core.generation

import org.example.core.Ticket

/**
 * This class is responsible for building ticket. It receives numbers via [place] function.
 *
 * After all numbers got distributed [build] function balances ticket's empty cells
 */
internal class TicketBuilder(val index: Int) {
    var numbersInTicketCount = 0
    private val columns = Array(9) { Column(it) }
    private val rows = Array(3) { arrayOfNulls<Int?>(9) }

    fun hasSpace(columnIndex: Int): Boolean {
        val howManyInColumn = columns[columnIndex].size
        return howManyInColumn < 3 && numbersInTicketCount < 15
    }

    /**
     * Places number in matching column. This function is not checking if number can be placed.
     *
     * It is expected that caller will use [hasSpace] function first or has other mechanism to track ticket and
     * column capacity
     */
    fun place(value: Int, columnIndex: Int) {
        numbersInTicketCount++
        columns[columnIndex].add(value)
    }

    fun build(): Ticket {
        this.balanceEmptyCells()
        return Ticket(rows)
    }

    private fun balanceEmptyCells() {
        val rowsEmptyCells = Array(3) { RowEmptyCells(it) }
        fun arrangeColumnWithTwoEmptyCells(column: Column){
            val rowIndicesLeft = mutableListOf(0, 1, 2)
            val ordered = rowsEmptyCells.sortedByDescending { c -> c.count }
            val columnIndex = column.index

            (0..1).map {
                val row = ordered[it]
                rowIndicesLeft.remove(row.index)
                row.count--
            }

            rows[rowIndicesLeft.first()][columnIndex] = column.first()
        }
        fun arrangeColumnWithOneEmptyCell(column: Column){
            val columnIndicesLeft = mutableListOf(0, 1, 2)

            val row = rowsEmptyCells.maxBy { c -> c.count }
            val columnIndex = column.index

            columnIndicesLeft.remove(row.index)
            row.count--

            val ordered = column.sortedBy { it }
            rows[columnIndicesLeft.first()][columnIndex] = ordered[0]
            rows[columnIndicesLeft.last()][columnIndex] = ordered[1]
        }
        fun arrangeColumnWithNoEmptyCells(column: Column){
            val columnIndex = column.index
            val ordered = column.sortedBy { it }

            rows[0][columnIndex] = ordered[0]
            rows[1][columnIndex] = ordered[1]
            rows[2][columnIndex] = ordered[2]
        }

        for (column in columns.sortedBy { c -> c.size }) {
            when (column.size){
                1 -> arrangeColumnWithTwoEmptyCells(column)
                2 -> arrangeColumnWithOneEmptyCell(column)
                3 -> arrangeColumnWithNoEmptyCells(column)
            }
        }
    }

    private class Column(val index: Int) : ArrayList<Int?>(3)
    private class RowEmptyCells(val index: Int){
        var count: Int = 4
    }
}